x-pack/plugins/secops/public/components/timeline/body/events/index.tsx

```diff
@@ -0,0 +1,177 @@
+                        <Actions
+                          associateNote={this.associateNote(event._id!, addNoteToEvent, onPinEvent)}
+                          expanded={!!this.state.expanded[event._id!]}
+                          data-test-subj="actions"
+                          eventId={event._id!}
+                          eventIsPinned={eventIsPinned({
+                            eventId: event._id!,
+                            pinnedEventIds,
+                          })}
+                          getNotesByIds={getNotesByIds}
+                          noteIds={eventIdToNoteIds[event._id!] || emptyNotes}
+                          onEventToggled={this.onToggleExpanded(event._id!)}
+                          onPinClicked={getPinOnClick({
+                            allowUnpinning: !eventHasNotes(eventIdToNoteIds[event._id!]),
+                            eventId: event._id!,
+                            onPinEvent,
+                            onUnPinEvent,
+                            pinnedEventIds,
+                          })}
+                          showNotes={!!this.state.showNotes[event._id!]}
+                          toggleShowNotes={this.onToggleShowNotes(event._id!)}
+                          updateNote={updateNote}
+                        />
+                      </EuiFlexItem>
+
+                      <EuiFlexItem grow={true}>
+                        <Columns
+                          data-test-subj="columns"
+                          columnHeaders={columnHeaders}
+                          columnRenderers={columnRenderers}
+                          ecs={event}
+                        />
+                        <NoteCards
+                          associateNote={this.associateNote(event._id!, addNoteToEvent, onPinEvent)}
+                          data-test-subj="note-cards"
+                          getNewNoteId={getNewNoteId}
+                          getNotesByIds={getNotesByIds}
+                          noteIds={eventIdToNoteIds[event._id!] || emptyNotes}
+                          showAddNote={!!this.state.showNotes[event._id!]}
+                          toggleShowAddNote={this.onToggleShowNotes(event._id!)}
+                          updateNote={updateNote}
+                          width="90%"
+                        />
+                      </EuiFlexItem>
+                    </EuiFlexGroup>
+                  </EuiFlexItem>
+
+                  <EuiFlexItem data-test-subj="event-details" grow={true}>
+                    <ExpandableEvent
+                      event={event}
+                      forceExpand={!!this.state.expanded[event._id!]}
+                      hideExpandButton={true}
+                      stringifiedEvent={stringifyEvent(event)}
+                      timelineId={id}
+                    />
+                  </EuiFlexItem>
+                </EuiFlexGroup>
+              </>
+            )}
+          </EuiFlexItem>
+        ))}
+      </EuiFlexGroup>
```

Comment:
This is kind of on the edge of being broken up into smaller functions reading through it. No changes required, just pointing out it's getting tricky auditing this function.

I also think that we need to really revisit having required attributes such as `_id!` when I see all the type safety being turned off, I don't know what the TypeScript is buying us other than either it should have been checked for `null` in the first place and we get a bug later during runtime or it should have never been null.

A good example is this chunk of code that I noticed below. We had ignored that `cursor` should be null with a `!` in a lot of places but then later we had bugs around it because as it turns out it _can_ be null. So then we added the null checks but left in the `!` which gives maintainers and other engineers confusion about when and where it can and cannot be null.

```ts
const beginning = cursor != null ? parseInt(cursor!, 10) : 0;
```

That code for example can now omit the bang as in:

```ts
const beginning = cursor != null ? parseInt(cursor, 10) : 0;
```

I would go the path of _making_ everything required and using TypeScript Partial when it needs to be optional or a large chunk of code needs to be optional and make functions take the Partial.

My 2 cents anyways. No changes needed here.

Reply:
> This is kind of on the edge of being broken up into smaller functions reading through it. No changes required, just pointing out it's getting tricky auditing this function with as large as it is.

Per our out-of-band video conversation regarding this function, we reviewed it and agreed that breaking this function up into smaller functions would result in a lot of extra parameter passing, but not much reduction in vertical space. (We also discussed that most of those components are taking functions that perform IO as parameters to facilitate unit testing, and we looked at one of those tests.)

We discussed (again, over video with @FrankHassanabad  and @spong) that the root cause of the `!`s here is the `_id` field in the `Ecs` `interface` being optional.

I created https://github.com/elastic/ingest-dev/issues/295 but in summary:

At present, the `Ecs` TS `interface`, which serves as the bridge between the GraphQL world and the rest of the UI, currently declares the `_id` field to be optional.

It's declared in `x-pack/plugins/secops/public/graphql/types.ts` as

```

export interface Ecs {
  _id?: string | null;
  // ...
}

```

As @FrankHassanabad has pointed out, the result of `_id` being optional is the proliferation of `!` throughout the codebase, with code like this:

```
<EuiFlexItem data-test-subj="event" grow={true} key={event._id!}>
```

...because in reality, we rely on `_id` being a valid, stable key in most contexts.

To resolve this issue:

1. The `Ecs` `interface` shall be updated, such that `_id` is no longer optional / nullable
2. All usages of `!` that were the result of it being optional shall be removed from the codebase

So no changes to the above code in this PR, but it makes sense to make `_id` non-optional, so we can remove the type assertions.